plan/aggregation_pruning.go

```diff
@@ -0,0 +1,123 @@
+// Copyright 2017 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+// // Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package plan
+
+import (
+	"github.com/juju/errors"
+	"github.com/pingcap/tidb/ast"
+	"github.com/pingcap/tidb/context"
+	"github.com/pingcap/tidb/expression"
+	"github.com/pingcap/tidb/mysql"
+	"github.com/pingcap/tidb/util/types"
+)
+
+type aggPruner struct {
+	allocator *idAllocator
+	ctx       context.Context
+}
+
+// eliminateAggregation will eliminate aggregation grouped by unique key.
+// e.g. select min(b) from t group by a. If a is a unique key, then this sql is equal to `select b from t group by a`.
+// For count(expr), sum(expr), avg(expr), we may need to rewrite the expr. Details is shown below.
+func (ap *aggPruner) eliminateAggregation(p LogicalPlan) error {
+	for _, child := range p.GetChildren() {
```

Comment:
It's not intuitive to process children. You can return a LogicalPlan instead and process p itself.